### PR TITLE
[DSCP-267] Forward-compatible serialisation prototype

### PR DESCRIPTION
**Pull request by Martoon-00**
### Description

Just prototype on how we can provide forward-compatible `Serialise` instances.
Ready for review.

### YT issue

https://issues.serokell.io/DSCP-267

### Introduced changes

- [x] New feature
- [ ] Bugfix 
- [ ] Refactoring
- [ ] New tests (on functionality not fixed/introduced in this PR)
- [ ] New docs (on functionality not fixed/introduced in this PR)
- [ ] Infrastructure changes 

### Checklist

If I added new functionality, I
- [ ] added tests covering it
- [ ] explained it using haddock in the source code

If I fixed a bug, I
- [ ] added a regression test to prevent the bug from silently reappearing again

If I changed public API structure or/and data serialization, I made sure that
- [ ] those changes are reflected in [Swagger API specs](/specs/disciplina)
- [ ] and also in [related](/docs/api-types.md) [documentation](/docs/authentication.md).

If I changed config structure or/and CLI interface of any executable, I made sure that
- [ ] [sample config](/docs/config-full-sample.yaml) is updated accordingly
- [ ] [launch scripts](/scripts/launch) are updated accordingly and work as expected

If I changed the procedure of building a project and/or running any executable or helper script, I
- [ ] updated [README.md](/README.md) accordingly
- [ ] as well as subproject READMEs for all related executables

Also,
- [ ] my commit history is clean, descriptive and do not contain merge or revert commits or commits like "WIP"
- [ ] my code complies with the [style guide](/docs/code-style.md)
- [ ] my documentation is checked with a spell checker (like [Grammarly](https://app.grammarly.com))

### CODE REVIEW

File: core/src/Dscp/Util/Serialise.hs [hunk 1]
```diff
@@ -50,3 +65,179 @@ decodeCrcProtected = do
...
+
+-- | Encode given complex data with remainder. You need only to 'encode'
+-- your datatype fields, length of list representing the type is filled for you.
+encodeWithRemainder :: Word -> Remainder -> E.Encoding -> E.Encoding
+encodeWithRemainder mainTermsN (Remainder terms) mainEncoder =
+    E.encodeListLen (mainTermsN + remainderSize) <>
+    mainEncoder <>
+    mconcat (map encodeTerm terms)
+  where
+    remainderSize = fromIntegral $ length terms :: Word
```
Comment (Heimdell) [inline core/src/Dscp/Util/Serialise.hs]: So, no "version" field?

File: core/src/Dscp/Util/Serialise.hs [hunk 2]
```diff
@@ -50,3 +65,179 @@ decodeCrcProtected = do
...
+instance NFData Term where
+    rnf = \case
+        TNull -> ()
+        _     -> ()
+
+---------------------------------------------------------------------------
+-- How will it look like
+---------------------------------------------------------------------------
+
+data Ololo = Ololo Int Double Remainder
```
Comment (Heimdell) [inline core/src/Dscp/Util/Serialise.hs]: Will real datatypes contain just plain `Remainder`? That doesn't look useful.

Comment (Heimdell) [reply core/src/Dscp/Util/Serialise.hs]: What _is_ the remainder? Later you just add fields before it, and that's it.

**Comment by Martoon-00** [comment]
I changed description a bit.

**Comment by Martoon-00** [comment]
I remembered about another requirement, that we need any node of a new version to survive a data of an old version stored in database, and my way does not help to fulfill it.

**Pull request closed without merging**